### ISSUE TITLE
<=> is not the same as ==

### ISSUE THREAD

**Issue by tdreyno** [hamstergem/hamster#246]
When using a `SortedSet`, there is a method called `uniq_by_comparator!`: https://github.com/hamstergem/hamster/blob/core/lib/hamster/sorted_set.rb#L88-L103

It assumes that items with the same sort position, `a <=> b == 0`, are actually the same object. That's not true. This causes `SortedSet`s with non-unique sort orders to not work correctly.

Imagine a comparator of `.size` where we are sorting on the length of some class. If the size is ever the same, they will be considered equal by Hamster.

For now, I'm working around this by adding `object_id` to my comparator, but that's definitely code smell:

```
      def <=>(other)
        [self.priority, self.object_id] <=> [other.priority, other.object_id]
      end
```

**Comment by alexdowad** [comment]
@tdreyno Thanks for your interest in the hamster gem, and I hope that it is
working well for you (aside from this issue).

What you are describing is the intended behavior of SortedSet. It is a
sorted "set", not a sorted "vector". The same comparator which defines the
sort order also defines what "equality" means for the set.

I thought the documentation was clear on that, but I wrote it years ago and
don't really remember exactly what it says. Please feel free to send a PR
if the documentation can be made more clear.

It seems that SortedSet is not the right structure for your use case.

**Comment by tdreyno** [comment]
Found a couple more things.

1. This only happens when adding items one at a time. If you add them all at once via the constructor, it works as expected.

2. The built-in Ruby SortedSet does not have this behavior. So the idea of overloading `<=>` to act as `==` in this implementation is at odds with the core library it is attempted to replace. 

Here is my test code to validate these claims:

```ruby
require 'set'
require 'hamster'

class Person
  include Comparable

  attr_reader :name
  attr_reader :age

  def <=>(other)
    age <=> other.age
  end

  def initialize(name, age)
    @name = name
    @age = age
  end
end

a = Person.new("a", 1)
b = Person.new("b", 5)
c = Person.new("c", 5)
d = Person.new("d", 10)

people = [d, c, b, a]

native_set = ::SortedSet.new
native_set << d
native_set << c
native_set << b
native_set << a

puts native_set.inspect, native_set.size

# Prints
# #<SortedSet: {#<Person:0x00007fede221a638 @name="a", @age=1>, #<Person:0x00007fede221a548 @name="c", @age=5>, #<Person:0x00007fede221a598 @name="b", @age=5>, #<Person:0x00007fede221a4d0 @name="d", @age=10>}>
# 4

hamster_set = ::Hamster::SortedSet.new
hamster_set = hamster_set << d
hamster_set = hamster_set << c
hamster_set = hamster_set << b
hamster_set = hamster_set << a

puts hamster_set.inspect, hamster_set.size

# Prints
# Hamster::SortedSet[#<Person:0x00007fede221a638 @name="a", @age=1>, #<Person:0x00007fede221a548 @name="c", @age=5>, #<Person:0x00007fede221a4d0 @name="d", @age=10>]
# 3

hamster_set2 = ::Hamster::SortedSet.new [d, c, b, a]

puts hamster_set2.inspect, hamster_set2.size

# Prints
# Hamster::SortedSet[#<Person:0x00007fede221a638 @name="a", @age=1>, #<Person:0x00007fede221a548 @name="c", @age=5>, #<Person:0x00007fede221a598 @name="b", @age=5>, #<Person:0x00007fede221a4d0 @name="d", @age=10>]
# 4
```

**Comment by alexdowad** [comment]
OK. Thanks for investigating further. Something is definitely wrong, but
it's not so straightforward to say what it is. There seems to be a conflict
between some of the library's basic values (as I see them).

One basic value is close compatibility with built-in Ruby classes like
Array. Maintaining this value would seem to mean that SortedSet's
documented interface should change to match Ruby's SortedSet.

Another value is that the immutable structures should be as performant as
possible. This presents a problem. The balanced tree structure used to
implement SortedSet will not work well if its semantics are changed as per
your wish. Almost all the important operations will become complicated to
implement, performance will be O(n) in the worst case rather than O(log n),
with a much worse constant factor...

And on top of this, I feel that the semantics of Hamster's SortedSet
actually make a lot of sense, once one understands them. Then there is also
backwards compatibility to think of.

As an aside, although I've not read the source code for Ruby's SortedSet,
from what you have shown me, I strongly suspect they are using a hash
table, with a doubly linked list of all hash nodes.

**Comment by alexdowad** [comment]
If it's not clear from the above, my preferred solution here is to fix the
SortedSet constructor so its behavior is consistent with repeated
insertion. But as mentioned, that would sacrifice some degree of
compatibility with the built-in one. A note to that effect would need to be
added to the docs.

**Comment by tdreyno** [comment]
The performance issue is a good one. I was actually digging through the code to figure out how it was fast, while constantly sorting.

What if, you made it an option. Give users a way of specifying Set equality (defaulting to `:==` for Set and `:<=>` for SortedSet), but with a hook for those of us who would prefer to match the core SortedSet library (but be immutable) and are willing to take the performance hit. My sets have only a couple hundred items AND I'm using them to avoid disk I/O which is way more costly for my use case.

That said, my hack of sorting on multiple parameters (using `object_id` to break the tie) also works.

**Comment by alexdowad** [comment]
Dear @tdreyno, your proposal is very interesting. Although the proposal
just involves adding to the API, rather than changing the existing API, I
would like to be very sure that this is the right thing to do. (Once
something is published to the world, it becomes difficult to change or
remove it, for fear of breaking other people's code.)

As for why `Hamster::SortedSet` is fast, see:
https://en.wikipedia.org/wiki/AVL_tree. Interestingly, while red-black
trees are a more commonly used type of balanced tree, and can easily be
implemented in an immutable, persistent manner, I found that AVL trees were
actually better and faster for this application. For a mutable data
structure, using red-black trees reduces the cost of modifying the
structure, since they don't need to be rebalanced as frequently. But for an
immutable structure, that doesn't help. Actually, it hurts, since the path
to the root needs to be copied on each modification, and maintaining strict
balance means the number of nodes copied for each such update is minimized.

Back to the original topic, it would be interesting and helpful to see an
implementation of your proposed structure. If the final decision is not to
add it to the library, you could release it as a gem which extends the
Hamster namespace with an additional class.

A simple (and not at all bad) implementation choice would be a class which
holds both a Hamster::Set and a Hamster::Vector. Use the Set for checking
inclusion, and the Vector for iteration and retrieving items by index. When
items are added or removed, update both the Set and the Vector.

Thanks!

**Comment by alexdowad** [comment]
I just opened up the Hamster source to see what was causing the problem with `SortedSet.new`... only to find a git branch with a bugfix for the same problem... which I created 6 months ago... (facepalm)

I'm trying to remember why this bugfix was never merged and released. I think that we have some hook in the Hamster repo which only allows merging dev branches if the tests pass on Travis CI... but Travis CI has become unreliable and often fails spuriously. If memory serves, I may have become frustrated trying to merge the branch with the fix and given up. Then again, the only one who can release new versions of the Hamster gem is @krainboltgreene, who hasn't been active here for a long time.

:(

I do also maintain a slightly cleaned-up fork of Hamster called `immutable-ruby`, which received the bugfix 6 months ago.
